### ISSUE TITLE
storage: cannot serve requested as_of 

### ISSUE THREAD

**Issue by sploiselle** [MaterializeInc/materialize#15426]
### What version of Materialize are you using?

 v0.27.0 (218138173)

### How did you install Materialize?

Materialize Cloud

### What is the issue?

1. Created a Kafka source with byte-formatted keys and values
3. Created a new cluster.
4. Created two replicas with differing sizes, `medium` and `xlarge`.
2. Created a materialized view to deserialized the bytes as JSON and then structured them.
3. Explicitly set my cluster to the new cluster.
4. Explicitly set my target replica to `medium`.
5. Queried the materialized view, which succeeded.
6. Explicitly set my target replica to `xlarge`.
7. Queried the materialized view, which never completed.
8. Checked logs in Honeycomb and determined that my worker had panicked:
    ```
    thread 'tokio-runtime-worker' panicked at 'u242: cannot serve requested as_of Antichain { elements: [1665768908791] }: Since(Antichain { elements: [1665769196001] })', /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0efb11835bec0ffd2-1/materialize/tests/src/storage/src/source/persist_source.rs:230:24
    ```

### Relevant log output

```source
Honeycomb produces the stack trace upside down--sorry!


compute-cluster-u8-replica-16-0	
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/blocking/pool.rs:300:13
compute-cluster-u8-replica-16-0	
  53: tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/blocking/pool.rs:325:17
compute-cluster-u8-replica-16-0	
  52: tokio::runtime::blocking::pool::Inner::run
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/blocking/pool.rs:91:9
compute-cluster-u8-replica-16-0	
  51: tokio::runtime::blocking::pool::Task::run
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/mod.rs:418:9
compute-cluster-u8-replica-16-0	
  50: tokio::runtime::task::UnownedTask<S>::run
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/raw.rs:84:18
compute-cluster-u8-replica-16-0	
  49: tokio::runtime::task::raw::RawTask::poll
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/harness.rs:57:15
compute-cluster-u8-replica-16-0	
  48: tokio::runtime::task::harness::Harness<T,S>::poll
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/harness.rs:104:27
compute-cluster-u8-replica-16-0	
  47: tokio::runtime::task::harness::Harness<T,S>::poll_inner
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/harness.rs:468:18
compute-cluster-u8-replica-16-0	
  46: tokio::runtime::task::harness::poll_future
compute-cluster-u8-replica-16-0	
             at ./rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panic.rs:137:14
compute-cluster-u8-replica-16-0	
  45: std::panic::catch_unwind
compute-cluster-u8-replica-16-0	
             at ./rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:456:19
compute-cluster-u8-replica-16-0	
  44: std::panicking::try
compute-cluster-u8-replica-16-0	
             at ./rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:492:40
compute-cluster-u8-replica-16-0	
  43: std::panicking::try::do_call
compute-cluster-u8-replica-16-0	
             at ./rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/panic/unwind_safe.rs:271:9
compute-cluster-u8-replica-16-0	
  42: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/harness.rs:480:19
compute-cluster-u8-replica-16-0	
  41: tokio::runtime::task::harness::poll_future::{{closure}}
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/core.rs:155:13
compute-cluster-u8-replica-16-0	
  40: tokio::runtime::task::core::CoreStage<T>::poll
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/loom/std/unsafe_cell.rs:14:9
compute-cluster-u8-replica-16-0	
  39: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/core.rs:165:17
compute-cluster-u8-replica-16-0	
  38: tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/blocking/task.rs:42:21
compute-cluster-u8-replica-16-0	
  37: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/thread_pool/worker.rs:362:45
compute-cluster-u8-replica-16-0	
  36: tokio::runtime::thread_pool::worker::Launch::launch::{{closure}}
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/thread_pool/worker.rs:383:5
compute-cluster-u8-replica-16-0	
  35: tokio::runtime::thread_pool::worker::run
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/macros/scoped_tls.rs:61:9
compute-cluster-u8-replica-16-0	
  34: tokio::macros::scoped_tls::ScopedKey<T>::set
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/thread_pool/worker.rs:386:17
compute-cluster-u8-replica-16-0	
  33: tokio::runtime::thread_pool::worker::run::{{closure}}
compute-cluster-u8-replica-16-0	
  32: tokio::runtime::thread_pool::worker::Context::run
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/thread_pool/worker.rs:434:9
compute-cluster-u8-replica-16-0	
  31: tokio::runtime::thread_pool::worker::Context::run_task
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/coop.rs:72:5
compute-cluster-u8-replica-16-0	
  30: tokio::coop::budget
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/coop.rs:95:5
compute-cluster-u8-replica-16-0	
  29: tokio::coop::with_budget
compute-cluster-u8-replica-16-0	
             at ./rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/thread/local.rs:421:9
compute-cluster-u8-replica-16-0	
  28: std::thread::local::LocalKey<T>::with
compute-cluster-u8-replica-16-0	
             at ./rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/thread/local.rs:445:16
compute-cluster-u8-replica-16-0	
  27: std::thread::local::LocalKey<T>::try_with
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/coop.rs:102:9
compute-cluster-u8-replica-16-0	
  26: tokio::coop::with_budget::{{closure}}
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/thread_pool/worker.rs:458:21
compute-cluster-u8-replica-16-0	
  25: tokio::runtime::thread_pool::worker::Context::run_task::{{closure}}
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/mod.rs:381:9
compute-cluster-u8-replica-16-0	
  24: tokio::runtime::task::LocalNotified<S>::run
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/raw.rs:84:18
compute-cluster-u8-replica-16-0	
  23: tokio::runtime::task::raw::RawTask::poll
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/harness.rs:57:15
compute-cluster-u8-replica-16-0	
  22: tokio::runtime::task::harness::Harness<T,S>::poll
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/harness.rs:104:27
compute-cluster-u8-replica-16-0	
  21: tokio::runtime::task::harness::Harness<T,S>::poll_inner
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/harness.rs:468:18
compute-cluster-u8-replica-16-0	
  20: tokio::runtime::task::harness::poll_future
compute-cluster-u8-replica-16-0	
             at ./rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panic.rs:137:14
compute-cluster-u8-replica-16-0	
  19: std::panic::catch_unwind
compute-cluster-u8-replica-16-0	
             at ./rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:456:19
compute-cluster-u8-replica-16-0	
  18: std::panicking::try
compute-cluster-u8-replica-16-0	
             at ./rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:492:40
compute-cluster-u8-replica-16-0	
  17: std::panicking::try::do_call
compute-cluster-u8-replica-16-0	
             at ./rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/panic/unwind_safe.rs:271:9
compute-cluster-u8-replica-16-0	
  16: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/harness.rs:480:19
compute-cluster-u8-replica-16-0	
  15: tokio::runtime::task::harness::poll_future::{{closure}}
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/core.rs:155:13
compute-cluster-u8-replica-16-0	
  14: tokio::runtime::task::core::CoreStage<T>::poll
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/loom/std/unsafe_cell.rs:14:9
compute-cluster-u8-replica-16-0	
  13: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.2/src/runtime/task/core.rs:165:17
compute-cluster-u8-replica-16-0	
  12: tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}
compute-cluster-u8-replica-16-0	
             at ./rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/future/mod.rs:91:19
compute-cluster-u8-replica-16-0	
  11: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
compute-cluster-u8-replica-16-0	
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0efb11835bec0ffd2-1/materialize/tests/src/storage/src/source/persist_source.rs:147:55
compute-cluster-u8-replica-16-0	
  10: <mz_storage::source::persist_source::DropSafeLeaseStream as core::ops::drop::Drop>::drop::{{closure}}
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.24/src/stream/stream/next.rs:32:9
compute-cluster-u8-replica-16-0	
   9: <futures_util::stream::stream::next::Next<St> as core::future::future::Future>::poll
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.24/src/stream/stream/mod.rs:1626:9
compute-cluster-u8-replica-16-0	
   8: futures_util::stream::stream::StreamExt::poll_next_unpin
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/futures-core-0.3.24/src/stream.rs:120:9
compute-cluster-u8-replica-16-0	
   7: <core::pin::Pin<P> as futures_core::stream::Stream>::poll_next
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/async-stream-0.3.3/src/async_stream.rs:53:17
compute-cluster-u8-replica-16-0	
   6: <async_stream::async_stream::AsyncStream<T,U> as futures_core::stream::Stream>::poll_next
compute-cluster-u8-replica-16-0	
             at ./rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/future/mod.rs:91:19
compute-cluster-u8-replica-16-0	
   5: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
compute-cluster-u8-replica-16-0	
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/async-stream-0.3.3/src/lib.rs:237:9
compute-cluster-u8-replica-16-0	
   4: mz_storage::source::persist_source::persist_source_core::{{closure}}
compute-cluster-u8-replica-16-0	
             at ./rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/result.rs:1493:23
compute-cluster-u8-replica-16-0	
   3: core::result::Result<T,E>::unwrap_or_else
compute-cluster-u8-replica-16-0	
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0efb11835bec0ffd2-1/materialize/tests/src/storage/src/source/persist_source.rs:230:24
compute-cluster-u8-replica-16-0	
   2: mz_storage::source::persist_source::persist_source_core::{{closure}}::{{closure}}
compute-cluster-u8-replica-16-0	
             at ./rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/panicking.rs:142:14
compute-cluster-u8-replica-16-0	
   1: core::panicking::panic_fmt
compute-cluster-u8-replica-16-0	
             at ./rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:584:5
compute-cluster-u8-replica-16-0	
   0: rust_begin_unwind
```
```

**Comment by philip-stoev** [comment]
If a reproduction scenario is needed:

1. Clone https://github.com/MaterializeInc/materialize/pull/15459/commits unless merged already
2. Run 
```
cd test/zippy
./mzcompose down -v 
./mzcompose run default --scenario=ClusterReplicas
```
3. While the test is running, fish for the panic in the logs:
```
$ docker logs zippy-materialized-1 2>&1 | grep 'panicked at' | grep as_of
thread 'timely:work-0' panicked at 'u96: cannot serve requested as_of Antichain { elements: [1665999786893] }: Since(Antichain { elements: [1665999788000] })', /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-07d043da2c32bb5e0-1/materialize/tests/src/storage/src/source/persist_source.rs:230:24
```

The test performs a lot of `CREATE CLUSTER REPLICA`, `DROP CLUSTER REPLICA` , killing for storageds and computeds and killing of individual replicas using `mz_panic()`

Note that the logs will contain other panics as well -- ones caused by Zippy using `mz_panic()` and those that occur due to the shared-fate mechanism.

**Comment by teskje** [comment]
### Reproduction

Thanks to Philip's instructions, I was able to come up to reproduce this locally. To do so, first apply this diff to the MZ source code, to ensure we always trigger the race condition:

```diff
diff --git a/src/compute/src/server.rs b/src/compute/src/server.rs
index 4a04a581d..bd2cebc60 100644
--- a/src/compute/src/server.rs
+++ b/src/compute/src/server.rs
@@ -98,6 +98,9 @@ pub type TimelyContainerRef = Arc<Mutex<Option<TimelyContainer>>>;
 pub fn serve(
     config: Config,
 ) -> Result<(TimelyContainerRef, impl Fn() -> Box<dyn ComputeClient>), Error> {
+    // simulate a slow replica startup
+    std::thread::sleep(std::time::Duration::from_secs(10));
+
     // Various metrics related things.
     let trace_metrics = TraceMetrics::register_with(&config.metrics_registry);

diff --git a/src/persist-client/src/read.rs b/src/persist-client/src/read.rs
index 06512fdbe..77c898101 100644
--- a/src/persist-client/src/read.rs
+++ b/src/persist-client/src/read.rs
@@ -677,15 +677,8 @@ where
     /// This is communicating actual progress information, so is given
     /// preferential treatment compared to Self::maybe_heartbeat_reader.
     pub async fn maybe_downgrade_since(&mut self, new_since: &Antichain<T>) {
-        // NB: min_elapsed is intentionally smaller than the one in
-        // maybe_heartbeat_reader (this is the preferential treatment mentioned
-        // above).
-        let min_elapsed = self.cfg.reader_lease_duration / 4;
-        let elapsed_since_last_heartbeat =
-            Duration::from_millis((self.cfg.now)().saturating_sub(self.last_heartbeat));
-        if elapsed_since_last_heartbeat >= min_elapsed {
-            self.downgrade_since(new_since).await;
-        }
+        // always downgrade immediately
+        self.downgrade_since(new_since).await;
     }

     /// Heartbeats the read lease if necessary.
```

Then perform the following steps:

1. Start environmentd: `bin/environmentd --reset`
2. Connect with psql 
3. Execute:
    ```sql
    CREATE TABLE t (a int);
    CREATE CLUSTER REPLICA default.r2 SIZE '1';
    SELECT * FROM t;
    ```
4. Wait for replica `r2` to come up after 10 seconds, and watch it crash.

### Cause

When the coordinator sequences a peek (i.e. SELECT) that cannot be served via the fast-path (i.e. there is no existing index maintained for it), it does the following:
* Install a transient dataflow that exports the required index
* Fire a peek command targeting that exported index
* Drop the transient dataflow again

Note that we don't wait for the peek to return before we drop its target dataflow. Instead we rely on the fact that the compute controller installs a read hold for each peek, that prevents dropping until the peek was served.

The compute controller considers a peek as served as soon as it receives the first peek response from any replica, or when it receives a `CancelPeeks` command. In both cases, it doesn't wait for all replicas to report that they have cancelled the peek and instead immediately allows the storage controller to compact the inputs of the transient dataflow. If all replicas have already managed to open their persist `ReadHandle`s by this time, we don't experience any issues because the `ReadHandle`s hold back compaction of the collections they read from. However, if the replica was slow to start rendering the dataflow (e.g. because it is an `xlarge` and took some time to start up), it will try to open a `ReadHandle` at a time that was already compacted away, and crash with the dreaded "cannot serve requested as_of" error.

**Comment by teskje** [comment]
> My current thinking is that if the compute controller kept track of the most behind replica (in addition to the most ahead replica), we could base our since holdback on that.

The compute controller actually already does that for the frontiers reported by replicas. It uses the upper bound of frontiers to track the global upper of the collection and the lower bound of frontiers to inform compaction decisions.

The main bug I think we currently have is that when we drop a dataflow, we immediately tell the storage controller that it is fine to compact it to the empty frontier, even before we have sent the `AllowCompaction` command to the replicas. So the compaction of the storage collection races with the replicas seeing the `AllowCompaction` command and stopping to read from those storage collections. This is currently fine in steady state, because `ReadHandle`s hold back compaction, but it will be not fine anymore when they stop doing that in the future.

**Comment by teskje** [comment]
I realized that #15504 does not completely solve this issue, because we also have replica-targeted peeks, for which we don't receive responses from all replicas. So even with this PR, the following still triggers the panic:

Apply the same steps as in the repro above, but this time run the following SQL statement:
```sql
CREATE TABLE t (a int);
CREATE CLUSTER REPLICA default.r2 SIZE '1';
SET cluster_replica = r1;
SELECT * FROM t;
```

The issue here is that replica r2 doesn't produce a peek response, so there is nothing the controller could wait on. But r2 still gets instructed to create the transient dataflow, and panics because that dataflow's input collection is already compacted when the replica gets around to rendering it.

I thought about adding a `PeekResponse::Ignored` to be emitted by replicas in response to peeks that don't target them. While that might be a good idea in general, it doesn't solve the issue at hand because the `PeekResponse::Ignored` can be emitted before the replica has created the peek dataflow, so the dataflow's source can still be compacted away underneath it.

I think to really solve this, we need to fix the bug I described in my previous comment, i.e. when dropping a dataflow, ensure that we allow compaction of the dataflow inputs only once all replicas have confirmed that they have stopped reading from them. I assume that this requires some refactoring of the compute controller and isn't as easy a change as #15504, so I think it's still a good idea to merge #15504 now, to get at least some relief.

**Comment by benesch** [comment]
> While that might be a good idea in general, it doesn't solve the issue at hand because the PeekResponse::Ignored can be emitted before the replica has created the peek dataflow, so the dataflow's source can still be compacted away underneath it.

Does this have to be true? Is it not possible to have the `PeekResponse::Ignored` only get emitted once the peek dataflow is built?

**Comment by teskje** [comment]
It is possible, we can make the decision as to whether to ignore a peek or not at the time we would serve the peek, rather than at the time we receive the peek. That would solve the bug, but it feels a bit icky because we are doing work we already know is unnecessary. We could try to implement it as a quick fix, but it is a bit scary to introduce a new compute response on short notice.

**Comment by benesch** [comment]
Sorry, I’m exposing my lack of familiarity with replica-targeted peeks here! Don’t we have to build identical dataflows on all replicas, even for targeted peeks, to get the identifiers to line up across workers?

**Comment by teskje** [comment]
No worries, I forgot that they existed as well, for some blissful time.

Yes, we build the dataflow everywhere, the only thing affected by the targeting is the handling of the peek command itself. So the dataflow processing work we need to do anyway. My assumption was that we also need some work for handling the peek itself in computed. But this is probably negligible and targeted peeks are not meant as a performance optimization anyway, so I can see that my argument is not quite valid.

It just occurred to me that we could handle the targeting inside `ActiveReplication` already, by ignoring all peek responses from untargeted replicas there. That would still allow us to track which replicas are finished with a peek, we wouldn't need a new compute response, and there would be one less edge-case to worry about in the rest of the client stack. This seems like a way better solution than what we currently have.

**Comment by benesch** [comment]
> It just occurred to me that we could handle the targeting inside ActiveReplication already, by ignoring all peek responses from untargeted replicas there. That would still allow us to track which replicas are finished with a peek, we wouldn't need a new compute response, and there would be one less edge-case to worry about in the rest of the client stack. This seems like a way better solution than what we currently have.

Indeed it does! 🙌🏽

**Comment by teskje** [comment]
Found another condition that triggers this panic. To repro, apply the same steps as in https://github.com/MaterializeInc/materialize/issues/15426#issuecomment-1282292997, but this time run:

```sql
CREATE TABLE t (a int);
DROP CLUSTER REPLICA default.r1;
CREATE CLUSTER REPLICA default.r1 SIZE '1';
SELECT * FROM t;
<ctrl-c>
```

This cancels the peek before the replica was ready to respond. The issue is that the compute controller removes the peek's read hold immediately when it receives the cancel command, it doesn't wait for the replicas to confirm that they are finished reading.

This is easy to fix, but I have a strong suspicion that this is still not the last cause for this panic.

**Comment by benesch** [comment]
🥳 Huge thank you, @teskje!

**Comment by danhhz** [comment]
> If you see any popping up with this changes, please scream!

got one of these on 0.30.1! https://sentry.io/organizations/materializeinc/issues/3678291063 Looks like the volume is way way down though, I can only find one on 0.30.1, the other recent ones I spot checked are 0.30.0-dev

**Comment by teskje** [comment]
After staring at the logs for some time I have a theory that requires validation by some storage folk (probably @danhhz):

In Sentry we currently see this panic occur in two customer envs. In both cases it happens in `mz_system`, which is a system cluster with a single replica and no standing dataflows. The only thing I imagine happens there is that the frontend regularly makes queries for objects installed in the system. Which would create an ad-hoc dataflow, peek it, then drop it. Because there is only a single replica, this is easy-mode for frontier tracking in the compute controller, so it would be surprising if the bug was there (not impossible though, and I'll get back to looking here if someone tells me that my first theory is unfounded).

I noticed that in both customer envs, environmentd crashed shortly after the compute replica crashed with the as_of panic. In both cases, environmentd panicked with:
```
thread 'tokio:work-0' panicked at 'ReaderId(...) was expired due to inactivity. Did the machine go to sleep?', /var/lib/buildkite-agent/builds/buildkite-bu
ilders-d43b1b5-i-07a25d21515e8928e-1/materialize/tests/src/persist-client/src/internal/state.rs:415:17
```

Probably an instance of this bug: https://github.com/MaterializeInc/materialize/issues/15209.

Now I wonder if this scenario is possible:
1. The compute controller correctly tracks replica frontiers and informs the storage controller about required read capabilities on some collection (let's say `mz_views`).
2. The storage controller's read handle on `mz_views` has its lease expired, losing its ability to hold back the shard's since.
3. The shard is compacted beyond the read capabilities the compute controller requested.
4. Compute replicas that relied on this read capability crash.
5. Environmentd crashes once the storage controller notices that it holds a read handle with an expired lease. 

Does this seem plausible?

**Comment by danhhz** [comment]
Yeah, that's almost certainly what happened, in fact. The good news is that "critical" read handles (coming ASAP) will fix this in a principled way #15511 #15822. The bad news is there isn't really a great mitigation until then.

As long as we recover from this on restart (thus don't get stuck in a crash loop), the volume we're seeing in practice probably means this is fine until we have critical handles. I _think_ we might have a real problem though, if this happens to an input of a dataflow that is being written to a materialized view.

**Comment by benesch** [comment]
I'll bet it was triggered by the Cilium networking trouble we were having, too. That should be more permanently fixed now, so hopefully we don't see this again in practice.
